src/content_scripts/autocorrect.js

```diff
@@ -0,0 +1,401 @@
+"use strict";
+
+const fractions = Object.freeze({
+    "¼": 1.0 / 4.0,
+    "½": 1.0 / 2.0,
+    "¾": 3.0 / 4.0,
+    "⅐": 1.0 / 7.0,
+    "⅑": 1.0 / 9.0,
+    "⅒": 1.0 / 10.0,
+    "⅓": 1.0 / 3.0,
+    "⅔": 2.0 / 3.0,
+    "⅕": 1.0 / 5.0,
+    "⅖": 2.0 / 5.0,
+    "⅗": 3.0 / 5.0,
+    "⅘": 4.0 / 5.0,
+    "⅙": 1.0 / 6.0,
+    "⅚": 5.0 / 6.0,
+    "⅛": 1.0 / 8.0,
+    "⅜": 3.0 / 8.0,
+    "⅝": 5.0 / 8.0,
+    "⅞": 7.0 / 8.0
+});
+
+const constants = Object.freeze({
+    "π": Math.PI,
+    "e": Math.E
+});
+
+// communication type
+// directly include magic constant as a workaround as we cannot import modules in content scripts due to https://bugzilla.mozilla.org/show_bug.cgi?id=1451545
+const AUTOCORRECT_CONTENT = "autocorrectContent";
+
+let insertedText; // Last insert text
+let deletedText; // Last deleted text
+let lastTarget; // Last target
+let lastCaretPosition; // Last caret position
+
+let quotes = true;
+let fracts = true;
+
+let autocorrections = {};
+
+let longest = 0;
+
+// Regular expressions
+let symbolpatterns = null;
+// Do not autocorrect for these patterns
+let antipatterns = null;
+
+// Thunderbird
+// https://bugzilla.mozilla.org/show_bug.cgi?id=1641573
+const THUNDERBIRD = typeof messenger !== "undefined";
+
+// Chrome
+// Adapted from: https://github.com/mozilla/webextension-polyfill/blob/master/src/browser-polyfill.js
+const CHROME = Object.getPrototypeOf(browser) !== Object.prototype;
+
+/**
+ * Get caret position.
+ *
+ * @param {Object} target
+ * @returns {number}
+ */
+function getCaretPosition(target) {
+    if (target.isContentEditable || THUNDERBIRD) {
+        target.focus();
+        const _range = document.getSelection().getRangeAt(0);
+        const range = _range.cloneRange();
+        const temp = document.createTextNode("\0");
+        range.insertNode(temp);
+        const caretposition = target.innerText.indexOf("\0");
+        temp.parentNode.removeChild(temp);
+        return caretposition;
+    } else {
+        return target.selectionStart;
+    }
+}
```

Comment:
Is this the best fallback? Can't we run `target.selectionStart` and test whether it returns `null` or `undefined` or so?

Reply:
The if clause is a hack to handle `ContentEditable` elements. The else clause is to handle `input` and `textarea` elements. It is supposed to return `undefined` when the caret is not in the target.

Reply:
Oh ok, I see, then could you please add these two "things it handles" as comments to each block? Thanks.